aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java

```diff
@@ -311,6 +314,88 @@ public synchronized void initialize(
+
+                if (result instanceof StorageResult.Success) {
+                    // If operation was successful, notify storage item change
+                    operation.getOnItemChange().accept(
+                            ((StorageResult.Success<T>) result).getStorageItemChange()
+                    );
+                } else if (result instanceof StorageResult.Failure && operation instanceof StorageOperation.Delete) {
+                    // If delete operation failed, assume record already deleted, 
+                    // so attempt publish storage item change
+                    try {
+                        final ModelSchema modelSchema =
+                                schemaRegistry.getModelSchemaForModelClass(operation.getModel().getModelName());
+                        StorageItemChange<T> change = StorageItemChange.<T>builder()
+                                .item(operation.getModel())
+                                .patchItem(SerializedModel.create(operation.getModel(), modelSchema))
+                                .modelSchema(modelSchema)
+                                .type(StorageItemChange.Type.DELETE)
+                                .predicate(QueryPredicates.all())
+                                .initiator(StorageItemChange.Initiator.SYNC_ENGINE)
+                                .build();
+                        operation.getOnItemChange().accept(change);
+                    } catch (Exception exception) {
+                        // do nothing
+                    }
+                } else if (result instanceof StorageResult.Failure) {
+                    DataStoreException exception = ((StorageResult.Failure<T>) result).getException();
+                    // If not a SQLiteConstraintException, we encountered an unrecoverable error
+                    // Return onError and don't proceed
+                    if (!ErrorInspector.contains(exception, SQLiteConstraintException.class)) {
+                        LOG.warn(
+                                "Failed to sync remote model into local storage: $modelWithMetadata",
+                                exception
+                        );
+                        // We still mark transaction as successful to save records that were already written
+                        sqlCommandProcessor.setTransactionSuccessful();
+                        sqlCommandProcessor.endTransaction();
+                        onError.accept(exception);
+                        return;
+                    } else {
+                        // LOG constraint violation errors and proceed
+                        LOG.warn(
+                                "Sync failed: foreign key constraint violation: $modelWithMetadata",
+                                exception
+                        );
+                    }
+                }
+            }
+            sqlCommandProcessor.setTransactionSuccessful();
+            sqlCommandProcessor.endTransaction();
```

Comment:
`endTransaction` should be in a `finally` block.

Reply:
good catch. Will need to be added before the other as well. Didn't realize `setTransactionSuccessful` can throw